src/pages/FrontPage.js

```diff
@@ -1,11 +1,214 @@
+      // Fix non-array tag lists.
+      // This happens if tags is created as string or numbers (e.g. someTagGroupName: tag1, tag2, 12345)
+      // instead of list (e.g. someTagGroupName: [tag1, tag2, 12345]) in YAML
+      if (typeof  tags === 'number') tags = tags.toString();
+      if (typeof tags === 'string') tags = tags.split(/,\s*/);
+
+      tagGroups.push({
+        name: groupName,
+        tags: tags
+      });
+    }
+
+    return tagGroups;
+  },
+  handleOnCheck(tag, checked) {
+    if(checked){
+      this.addFilter(tag);
+    }else{
+      this.removeFilter(tag);
+    }
+  },
+  addFilter(tag) {
+    let filter = this.state.filter;
+    filter.push({name: tag.groupName, tags: [tag.name]});
+    // Merge tagGroups with same name
+    filter = this.getListWithDistinctObjects(filter, this.mergeTagGroups);
+
+    const filteredCourses = this.filterCourses(this.state.allCourses, filter);
+    this.setState({
+      filter,
+      filteredCourses
+    });
+  },
+  removeFilter(tag) {
+    const filter = this.state.filter;
+
+    // Find the tagGroup and remove tag from it
+    const tagGroupFound = filter.find(tagGroup => tagGroup.name === tag.groupName);
+    if(tagGroupFound) {
+      const tags = tagGroupFound.tags;
+      if(tags.length < 2) {
+        // Remove the tagGroup if it only has one tag
+        filter.splice(filter.indexOf(tagGroupFound), 1);
+      } else if(tags.indexOf(tag.name) >= 0){
+        // Remove tag from tagGroup
+        tags.splice(tags.indexOf(tag.name), 1);
+      }
+    }
+    const filteredCourses = this.filterCourses(this.state.allCourses, filter);
+    this.setState({
+      filteredCourses
+    });
+  },
+  filterCourses(courses, filter) {
+    // Clone all courses to prevent side-effects
+    courses = JSON.parse(JSON.stringify(courses));
+
+    // Find and sort courses that have at least one lesson that matches filter
+    return courses.filter((course) => {
+      // Remove lessons from course that doesn't match filter
+      course.lessons = this.filterLessons(course.lessons, filter);
+      return course.lessons.length > 0;
+    }).sort(this.sortCourses);
+
+  },
+  filterLessons(lessons, filter) {
+    // Find lessons that matches filter
+    return lessons.filter((lesson) => {
+      for(var i = 0; i < filter.length; i++) {
+        const filterTagGroup = filter[i];
+        if( !this.lessonHasTag(lesson, filterTagGroup) )return false;
+      }
+      return true;
+    });
+  },
+  lessonHasTag(lesson, filterTagGroup) {
+    const lessonTagGroups = lesson.tagGroups;
+    if(lessonTagGroups.length == 0 && filterTagGroup !== null) return false;
+    for(var i = 0; i < lessonTagGroups.length; i++){
+      const lessonTagGroup = lessonTagGroups[i];
+      //TODO: Clean
+      if(lessonTagGroup.name.toLowerCase() == filterTagGroup.name.toLowerCase() && this.isSubArray(filterTagGroup.tags, lessonTagGroup.tags))return true;
+    }
+    return false;
+  },
+  mergeTagGroups(tagGroupA, tagGroupB) {
```

Comment:
Denne funksjonen kan du skrive litt enklere:
```
tagGroupA.tags = [...new Set(tagGroupA.tags.concat(tagGroupB.tags))];
```

Her bruker jeg trikset med at
```
const uniqueArray  = [ ...new Set(array) ]
```

Når det er sagt, så prøv å unngå funksjoner som endrer argumentene, altså prøv å lage "pure" functions, uten side effects. Mye lettere å teste, mye mindre sjanse for feil.